### PR TITLE
Make generateKeyFrame take a single RID or none. Resolve the promise with the frame timestamp

### PR DESCRIPTION
**Pull request by youennf**
Fixes https://github.com/w3c/webrtc-encoded-transform/issues/128

***
<a href="https://pr-preview.s3.amazonaws.com/youennf/webrtc-insertable-streams/pull/132.html" title="Last updated on May 19, 2022, 2:40 PM UTC (d33d388)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/webrtc-encoded-transform/132/da20661...youennf:d33d388.html" title="Last updated on May 19, 2022, 2:40 PM UTC (d33d388)">Diff</a>

### CODE REVIEW

**Comment by youennf** [comment]
@aboba, @alvestrand, thoughts?
This basically remove the ability to ask for key frame for all encoders, which then allows to expose the timestamp of the generated key frame when resolving the promise.

**Comment by fippo** [comment]
how do you plan on implementing this? The rationale for generating keyframes for all simulcast streams at the same time given in https://bugs.chromium.org/p/webrtc/issues/detail?id=10107#c4 was pretty clear. Not that I agree with it but...

**Comment by youennf** [comment]
> how do you plan on implementing this?

This is a problem with or without these changes.
The spec allows implementations to trigger a key frame to all related encoders when only one is actually requested.

If we wanted to go with https://bugs.chromium.org/p/webrtc/issues/detail?id=10107#c4, I think we would then remove the rid parameter entirely. Given this API is based on an initial Google proposal that was already exposing the rid parameter, there might be a possibility to fix https://bugs.chromium.org/p/webrtc/issues/detail?id=10107, at least in some cases. @alvestrand thoughts?

**Comment by youennf** [comment]
This was discussed at last WebRTC interim and reached rough consensus at that time.

File: index.bs [hunk 1]
```diff
@@ -408,23 +408,24 @@ The <dfn attribute for="RTCRtpScriptTransformer">writable</dfn> getter steps are
...
-        1. If |encoder|.`[[pendingKeyFrameTasks]]` contains a task which is matching |rid|, remove |rid| from |rids|.
-        1. Add |task| to |encoder|.`[[pendingKeyFrameTasks]]`.
-    1. If |rids| is not empty, remove from |videoEncoders| any video encoder that does not match a value in |rids|.
-    1. For each |videoEncoder| in |videoEncoders|, instruct |videoEncoder| to generate a key frame for the next provided video frame.
+    1. Let |videoEncoder| be the first encoder in |videoEncoders|.
+    1. If |rid| is undefined, set |rid| to the RID value corresponding to |videoEncoder|.
+    1. Create a pending key frame task called |task| with |task|.`[[rid]]` set to rid and |task|.`[[promise]]`| set to |promise|.
+    1. If |encoder|.`[[pendingKeyFrameTasks]]` is undefined, initialize |encoder|.`[[pendingKeyFrameTasks]]` to an empty set.
+    1. Let |shouldTriggerKeyFrame| be <code>true</code> if |encoder|.`[[pendingKeyFrameTasks]]` contains a task whose `[[rid]]`
+        value which is equal to |rid|, and <code>false</code> otherwise.
```
Comment (youennf) [inline index.bs:426]:
```suggestion
        value is equal to |rid|, and <code>false</code> otherwise.
```

**Pull request merged**